### PR TITLE
posix/unix: handle correctly removal of connected socket

### PR DESCRIPTION
**Pull request by ziemleszcz**
- set peer socket to disconnected state
- send on disconnected socket returns EPIPE (SOCK_STREAM, SOCK_SEQPACKET)
- send on disconnected socket return ECONNREFUSED (SOCK_DGRAM)
- send on disconnected socket sends SIGPIPE (SOCK_STREAM, SOCKSEQPACKET)
- recv on disconnected sockets returns 0 (SOCK_STREAM, SOCK_SEQPACKET)

JIRA: DTR-241

## Motivation and Context

Required for openiked. Currently send/recv on socket connected to closed socket will crash kernel.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: posix/unix.c [hunk 1]
```diff
@@ -218,20 +218,26 @@ static unixsock_t *unixsock_get(unsigned id)
...
 	proc_lockSet(&unix_common.lock);
-	if (--r->refs <= 0) {
-		lib_rbRemove(&unix_common.tree, &r->linkage);
+	if (--s->refs <= 0) {
+		lib_rbRemove(&unix_common.tree, &s->linkage);
+
+		/* TODO: handle connecting socket */
+
+		if (s->connect != NULL && !(s->state & US_DISCONNECTED))
+			s->connect->state |= US_DISCONNECTED;
```
Comment (nalajcie) [inline posix/unix.c:230]: maybe we should set s->connect->connect = NULL  to avoid dangling pointers?

File: posix/unix.c [hunk 2]
```diff
@@ -698,6 +713,11 @@ static ssize_t send(unsigned socket, const void *buf, size_t len, int flags, con
 				err = -ENOTCONN;
 				break;
 			}
+			else if (s->state & US_DISCONNECTED) {
```
Comment (nalajcie) [inline posix/unix.c:716]: not sure, but it seems that we will be sending `SIGPIPE` on every `send()`, maybe we should instead send it once and set `s->connect = NULL` here and and clear `US_DISCONNECTED` flag?

File: posix/unix.c [hunk 3]
```diff
@@ -26,11 +26,11 @@
...
-#define US_BOUND      (1 << 0)
-#define US_LISTENING  (1 << 1)
-#define US_ACCEPTING  (1 << 2)
-#define US_CONNECTING (1 << 3)
-
+#define US_BOUND        (1 << 0)
+#define US_LISTENING    (1 << 1)
+#define US_ACCEPTING    (1 << 2)
+#define US_CONNECTING   (1 << 3)
+#define US_DISCONNECTED (1 << 4) /* NOTE: invalidates connect pointer */
```
Comment (nalajcie) [inline posix/unix.c:33]: Hm... new socket is not in `disconnected` state, it's a special flag that indicates the socket was already connected and than disconnected from the other side (sooooo.... it's "disconnecting" now as "disconnected" sockets have state == 0?)

File: posix/unix.c [hunk 4]
```diff
@@ -373,6 +379,7 @@ int unix_accept4(unsigned socket, struct sockaddr *address, socklen_t *address_l
 		LIST_REMOVE(&s->connect, conn);
 		s->state &= ~US_ACCEPTING;
 
+		conn->state &= ~US_DISCONNECTED;
```
Comment (nalajcie) [inline posix/unix.c:382]: maybe only clearing after sending SIGPIPE (+ add sigpipe sending to all socket operations?) to reset to "disconnected" state?

**Review by nalajcie — COMMENTED** [review]
Not sure about full unix socket semantics, so leaving only comments ;)

File: posix/unix.c (hunk 1 above)
Comment (ziemleszcz) [reply posix/unix.c:230]: Yeah, dangling pointer is not nice.

File: posix/unix.c (hunk 2 above)
Comment (ziemleszcz) [reply posix/unix.c:716]: Linux sends SIGPIPE on every send.

File: posix/unix.c (hunk 3 above)
Comment (ziemleszcz) [reply posix/unix.c:33]: It should really be US_PEER_DISCONNECTED but I've decided it's too long.

File: posix/unix.c (hunk 4 above)
Comment (ziemleszcz) [reply posix/unix.c:382]: I don't think SIGPIPE should change socket state, it's related only to send() call (based on POSIX spec).

Comment (nalajcie) [reply posix/unix.c:382]: ok, it was just an idea to send `SIGPIPE` only once, if not needed - let's drop it.

Comment (ziemleszcz) [reply posix/unix.c:382]: Turns out Linux forbids reconnecting stream socket, connect() call on peer disconnected socket return EISCONN.

**Review by agkaminski — APPROVED** [review]
LGTM

**Comment by ziemleszcz** [comment]
Hey, not so fast. I marked that I will merge it myself. It was draft commit.

**Pull request merged**